core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java

```diff
@@ -53,20 +53,22 @@ public class ResolvingFileIO implements HadoopConfigurable, DelegateFileIO {
   private static final String S3_FILE_IO_IMPL = "org.apache.iceberg.aws.s3.S3FileIO";
   private static final String GCS_FILE_IO_IMPL = "org.apache.iceberg.gcp.gcs.GCSFileIO";
   private static final String ADLS_FILE_IO_IMPL = "org.apache.iceberg.azure.adlsv2.ADLSFileIO";
-  private static final Map<String, String> SCHEME_TO_FILE_IO =
+  private static final Map<String, String> DEFAULT_SCHEME_TO_FILE_IO =
       ImmutableMap.of(
           "s3", S3_FILE_IO_IMPL,
           "s3a", S3_FILE_IO_IMPL,
           "s3n", S3_FILE_IO_IMPL,
           "gs", GCS_FILE_IO_IMPL,
           "abfs", ADLS_FILE_IO_IMPL,
           "abfss", ADLS_FILE_IO_IMPL);
+  private static final String SCHEME_PROPERTY_PREFIX = "resolving-io.schemes.";
 
   private final Map<String, DelegateFileIO> ioInstances = Maps.newConcurrentMap();
   private final AtomicBoolean isClosed = new AtomicBoolean(false);
   private final transient StackTraceElement[] createStack;
   private SerializableMap<String, String> properties;
   private SerializableSupplier<Configuration> hadoopConf;
+  private final Map<String, String> schemeToFileIo = Maps.newHashMap();
```

Comment:
Maybe switch to `Maps.newConcurrentMap()`?

Reply:
I don't think a `ConcurrentMap` is required here.  If so, we need to change the `scheme()` method as `ConcurrentMap` will throw `NullPointerException` for `getOrDefault()` if the specified key is null and this map does not permit null keys.

Reply:
@ConeyLiu has changed `ioInstances` to be a concurrent map in 9ec6084cc21217dc27bb5db870fb0d7b607e3ad1, so I suppose there is a reason for that, and that this class is supposed to be accessed concurrently.

If this assumption is correct, I think it's important to use a concurrent map here as well.

Reply:
I can understand for `ioInstances`. But you are right, let "secure" the scheme handlers 😄 

I gonna change `scheme()` method to work with `ConcurrentMap` then.